README.md
# Flow.js [![Build Status](https://travis-ci.org/flowjs/flow.js.svg)](https://travis-ci.org/flowjs/flow.js) [![Test Coverage](https://codeclimate.com/github/flowjs/flow.js/badges/coverage.svg)](https://codeclimate.com/github/flowjs/flow.js/coverage) [![Saucelabs Test Status](https://saucelabs.com/browser-matrix/flowjs.svg)](https://saucelabs.com/u/flowjs) [![Inline docs](http://inch-ci.org/github/flowjs/flow.js.svg?branch=master)](http://inch-ci.org/github/flowjs/flow.js)

<a href="https://www.buymeacoffee.com/aidas" target="_blank"><img src="https://cdn.buymeacoffee.com/buttons/default-orange.png" alt="Buy Me A Coffee" style="height: 51px !important;width: 217px !important;" ></a>

Flow.js is a JavaScript library providing multiple simultaneous, stable and resumable uploads via the HTML5 File API. 

The library is designed to introduce fault-tolerance into the upload of large files through HTTP. This is done by splitting each file into small chunks. Then, whenever the upload of a chunk fails, uploading is retried until the procedure completes. This allows uploads to automatically resume uploading after a network connection is lost either locally or to the server. Additionally, it allows for users to pause, resume and even recover uploads without losing state because only the currently uploading chunks will be aborted, not the entire upload.

Flow.js does not have any external dependencies other than the `HTML5 File API`. This is relied on for the ability to chunk files into smaller pieces. Currently, this means that support is limited to Firefox 4+, Chrome 11+, Safari 6+ and Internet Explorer 10+.

Samples and examples are available in the `samples/` folder. Please push your own as Markdown to help document the project.

## Can I see a demo?
[Flow.js + angular.js file upload demo](http://flowjs.github.io/ng-flow/) - ng-flow extension page https://github.com/flowjs/ng-flow

JQuery and node.js backend demo https://github.com/flowjs/flow.js/tree/master/samples/Node.js

## How can I install it?

Download a latest build from https://github.com/flowjs/flow.js/releases
it contains development and minified production files in `dist/` folder.

or use npm:
```console
npm install @flowjs/flow.js
```

or use bower:
```console
bower install flow.js#~2
```                
or use git clone
```console
git clone https://github.com/flowjs/flow.js
```
## How can I use it?

A new `Flow` object is created with information of what and where to post:
```javascript
var flow = new Flow({
  target:'/api/photo/redeem-upload-token', 
  query:{upload_token:'my_token'}
});
// Flow.js isn't supported, fall back on a different method
if(!flow.support) location.href = '/some-old-crappy-uploader';
```
To allow files to be either selected and drag-dropped, you'll assign drop target and a DOM item to be clicked for browsing:
```javascript
flow.assignBrowse(document.getElementById('browseButton'));
flow.assignDrop(document.getElementById('dropTarget'));
```
After this, interaction with Flow.js is done by listening to events:
```javascript
flow.on('fileAdded', function(file, event){
    console.log(file, event);
});
flow.on('fileSuccess', function(file,message){
    console.log(file,message);
});
flow.on('fileError', function(file, message){
    console.log(file, message);
});
```
## How do I set it up with my server?

Most of the magic for Flow.js happens in the user's browser, but files still need to be reassembled from chunks on the server side. This should be a fairly simple task and can be achieved in any web framework or language, which is able to receive file uploads.

To handle the state of upload chunks, a number of extra parameters are sent along with all requests:

* `flowChunkNumber`: The index of the chunk in the current upload. First chunk is `1` (no base-0 counting here).
* `flowTotalChunks`: The total number of chunks.  
* `flowChunkSize`: The general chunk size. Using this value and `flowTotalSize` you can calculate the total number of chunks. Please note that the size of the data received in the HTTP might be lower than `flowChunkSize` of this for the last chunk for a file.
* `flowTotalSize`: The total file size.
* `flowIdentifier`: A unique identifier for the file contained in the request.
* `flowFilename`: The original file name (since a bug in Firefox results in the file name not being transmitted in chunk multipart posts).
* `flowRelativePath`: The file's relative path when selecting a directory (defaults to file name in all browsers except Chrome).

You should allow for the same chunk to be uploaded more than once; this isn't standard behaviour, but on an unstable network environment it could happen, and this case is exactly what Flow.js is designed for.

For every request, you can confirm reception in HTTP status codes (can be change through the `permanentErrors` option):

* `200`, `201`, `202`: The chunk was accepted and correct. No need to re-upload.
* `404`, `415`. `500`, `501`: The file for which the chunk was uploaded is not supported, cancel the entire upload.
* _Anything else_: Something went wrong, but try reuploading the file.

## Handling GET (or `test()` requests)

Enabling the `testChunks` option will allow uploads to be resumed after browser restarts and even across browsers (in theory you could even run the same file upload across multiple tabs or different browsers).  The `POST` data requests listed are required to use Flow.js to receive data, but you can extend support by implementing a corresponding `GET` request with the same parameters:

* If this request returns a `200`, `201` or `202` HTTP code, the chunks is assumed to have been completed.
* If request returns a permanent error status, upload is stopped.
* If request returns anything else, the chunk will be uploaded in the standard fashion.

After this is done and `testChunks` enabled, an upload can quickly catch up even after a browser restart by simply verifying already uploaded chunks that do not need to be uploaded again.

## Full documentation

### Flow
#### Configuration

The object is loaded with a configuration options:
```javascript
var r = new Flow({opt1:'val', ...});
```
Available configuration options are:

* `target` The target URL for the multipart POST request. This can be a string or a function. If a
function, it will be passed a FlowFile, a FlowChunk and isTest boolean (Default: `/`)
* `singleFile` Enable single file upload. Once one file is uploaded, second file will overtake existing one, first one will be canceled. (Default: false)
* `chunkSize` The size in bytes of each uploaded chunk of data. This can be a number or a function. If a function, it will be passed a FlowFile. The last uploaded chunk will be at least this size and up to two the size, see [Issue #51](https://github.com/23/resumable.js/issues/51) for details and reasons. (Default: `1*1024*1024`, 1MB)
* `forceChunkSize` Force all chunks to be less or equal than chunkSize. Otherwise, the last chunk will be greater than or equal to `chunkSize`. (Default: `false`)
* `simultaneousUploads` Number of simultaneous uploads (Default: `3`)
* `fileParameterName` The name of the multipart POST parameter to use for the file chunk  (Default: `file`)
* `query` Extra parameters to include in the multipart POST with data. This can be an object or a
 function. If a function, it will be passed a FlowFile, a FlowChunk object and a isTest boolean
 (Default: `{}`)
* `headers` Extra headers to include in the multipart POST with data. If a function, it will be passed a FlowFile, a FlowChunk object and a isTest boolean (Default: `{}`)
* `withCredentials` Standard CORS requests do not send or set any cookies by default. In order to
 include cookies as part of the request, you need to set the `withCredentials` property to true.
(Default: `false`)
* `method` Method to use when POSTing chunks to the server (`multipart` or `octet`) (Default: `multipart`)
* `testMethod` HTTP method to use when chunks are being tested. If set to a function, it will be passed a FlowFile and a FlowChunk arguments. (Default: `GET`)
* `uploadMethod` HTTP method to use when chunks are being uploaded. If set to a function, it will be passed a FlowFile and a FlowChunk arguments. (Default: `POST`)
* `allowDuplicateUploads ` Once a file is uploaded, allow reupload of the same file. By default, if a file is already uploaded, it will be skipped unless the file is removed from the existing Flow object. (Default: `false`)
* `prioritizeFirstAndLastChunk` Prioritize first and last chunks of all files. This can be handy if you can determine if a file is valid for your service from only the first or last chunk. For example, photo or video meta data is usually located in the first part of a file, making it easy to test support from only the first chunk. (Default: `false`)
* `testChunks` Make a GET request to the server for each chunks to see if it already exists. If implemented on the server-side, this will allow for upload resumes even after a browser crash or even a computer restart. (Default: `true`)
* `preprocess` Optional function to process each chunk before testing & sending. To the function it will be passed the chunk as parameter, and should call the `preprocessFinished` method on the chunk when finished. (Default: `null`)
* `changeRawDataBeforeSend` Optional function to change Raw Data just before the XHR Request can be sent for each chunk. To the function, it will be passed the chunk and the data as a Parameter. Return the data which will be then sent to the XHR request without further modification. (Default: `null`). This is helpful when using FlowJS with [Google Cloud Storage](https://cloud.google.com/storage/docs/json_api/v1/how-tos/multipart-upload). Usage example can be seen [#276](https://github.com/flowjs/flow.js/pull/276). (For more, check issue [#170](https://github.com/flowjs/flow.js/issues/170)).
* `initFileFn` Optional function to initialize the fileObject. To the function it will be passed a FlowFile and a FlowChunk arguments.
* `readFileFn` Optional function wrapping reading operation from the original file. To the function it will be passed the FlowFile, the startByte and endByte, the fileType and the FlowChunk.
* `generateUniqueIdentifier` Override the function that generates unique identifiers for each file. (Default: `null`)
* `maxChunkRetries` The maximum number of retries for a chunk before the upload is failed. Valid values are any positive integer and `undefined` for no limit. (Default: `0`)
* `chunkRetryInterval` The number of milliseconds to wait before retrying a chunk on a non-permanent error.  Valid values are any positive integer and `undefined` for immediate retry. (Default: `undefined`)
* `progressCallbacksInterval` The time interval in milliseconds between progress reports. Set it
to 0 to handle each progress callback. (Default: `500`)
* `speedSmoothingFactor` Used for calculating average upload speed. Number from 1 to 0. Set to 1
and average upload speed wil be equal to current upload speed. For longer file uploads it is
better set this number to 0.02, because time remaining estimation will be more accurate. This
parameter must be adjusted together with `progressCallbacksInterval` parameter. (Default 0.1)
* `successStatuses` Response is success if response status is in this list (Default: `[200,201,
202]`)
* `permanentErrors` Response fails if response status is in this list (Default: `[404, 415, 500, 501]`)


#### Properties

* `.support` A boolean value indicator whether or not Flow.js is supported by the current browser.
* `.supportDirectory` A boolean value, which indicates if browser supports directory uploads.
* `.opts` A hash object of the configuration of the Flow.js instance.
* `.files` An array of `FlowFile` file objects added by the user (see full docs for this object type below).

#### Methods


CHANGELOG.md
# 3.0.0

## Breaking Changes

### Events
 - Events are distincts from hooks.
 - Recognized events and hooks are now lower-snake-case (case-sensitive).
 - Hooks are "file-added", "files-added", "files-submitted". "filter-file" is a filtering hook.
 - Events are passed a native CustomEvent. IE:
    v2: `flow.on('fileRemoved',  (file) => { ... });`
    v3: `flow.on('file-removed', ({detail: [file]}) => { ... });`

### Other
 - FlowFile does not run bootstrap() upon instanciation. This must be done manually (or rely on *addFile* functions).

# 2.0.0

## Features

 - All code follows Google javascript style guide
 - Target url can be provided with query string
 - Events **fileAdded** and **filesAdded** can prevent file from being added to $.files list by
 returning false. Custom validators can be ran here.
 - **ResumableFile.getType()** and **ResumableFile.getExtension()** helper methods added. Can be
 used for custom validation.
 - **fileProgress** and **progress** events are always asynchronous.
 - **ResumableFile.pause()** and **ResumableFile.resume()** methods for single file pausing and
 resuming.
 - **filesSubmitted** event added. Can be used to start file upload. Event is thrown then files are
 added to queue.
 - **progressCallbacksInterval** parameter added. Minimum interval between callbacks execution in
  milliseconds.
 - **averageSpeed** and **currentSpeed** parameters added for `ResumableFile`. These params
 accuracy can be adjusted with `speedSmoothingFactor` and `progressCallbacksInterval` parameters.
 - **timeRemaining** method added for `ResumableFile`. Returns remaining time to upload in seconds. Accuracy is based on average speed.
 - **sizeUploaded** method added for `ResumableFile`. Returns size uploaded in bytes.
 - **singleFile** parameter added. Then enabled, uploaded file will replace current one.

## Breaking Changes
 - **Resumable** was renamed to **Flow**
 - **ResumableFile.fileName** parameter renamed to **ResumableFile.name**
 - **Resumable.getOpt** method dropped, use Resumable.opts parameter instead if needed.
 - **Resumable.maxFiles**, **Resumable.minFileSize**, **Resumable.maxFileSize**,
 **Resumable.fileType** validators dropped. Use **fileAdded** and **filesAdded** events for
 custom validation.
 - **fileProgress** and **progress** events are not thrown on ResumableFile.abort() and ResumableFile.cancel() methods execution.
 - **cancel** event was removed. Event was always called after **Resumable.cancel()** function.
 - **fileAdded**, **filesAdded** events are thrown before file is added to upload queue. This means
 that calling **Resumable.upload()** method in these events will not start uploading current
 files. To start upload use **filesSubmitted** event instead.
 - **throttleProgressCallbacks** parameter was replaced with **progressCallbacksInterval** and it
  is now measured in milliseconds.


src/Flow.js
/**
 * @license MIT
 */

import Eventizer from './Eventizer';
import FlowFile from './FlowFile';
import AsyncFlowFile from './AsyncFlowFile';
import {each, webAPIFileRead} from './tools';

/**
 * Flow.js is a library providing multiple simultaneous, stable and
 * resumable uploads via the HTML5 File API.
 * @param [opts]
 * @param {number|Function} [opts.chunkSize]
 * @param {bool} [opts.forceChunkSize]
 * @param {number} [opts.simultaneousUploads]
 * @param {bool} [opts.singleFile]
 * @param {string} [opts.fileParameterName]
 * @param {number} [opts.progressCallbacksInterval]
 * @param {number} [opts.speedSmoothingFactor]
 * @param {Object|Function} [opts.query]
 * @param {Object|Function} [opts.headers]
 * @param {bool} [opts.withCredentials]
 * @param {Function} [opts.preprocess]
 * @param {string} [opts.method]
 * @param {string|Function} [opts.testMethod]
 * @param {string|Function} [opts.uploadMethod]
 * @param {bool} [opts.prioritizeFirstAndLastChunk]
 * @param {bool} [opts.allowDuplicateUploads]
 * @param {string|Function} [opts.target]
 * @param {number} [opts.maxChunkRetries]
 * @param {number} [opts.chunkRetryInterval]
 * @param {Array.<number>} [opts.permanentErrors]
 * @param {Array.<number>} [opts.successStatuses]
 * @param {Function} [opts.initFileFn]
 * @param {Function} [opts.readFileFn]
 * @param {Function} [opts.asyncReadFileFn]
 * @param {Function} [opts.generateUniqueIdentifier]
 * @constructor
 */
export default class Flow extends Eventizer {

  /**
   * For the events object:
   *  - keys: stands for event name
   *  - values: array list of callbacks
   * All keys are lowercased as on() would do.
   */
  constructor(opts, events) {
    super(events);

    /**
     * Library version
     * @type {string}
     */
    Flow.version = '<%= version %>';

    /**
     * Check if directory upload is supported
     * @type {boolean}
     */
    this.supportDirectory = (
      /Chrome/.test(window.navigator.userAgent) ||
      /Firefox/.test(window.navigator.userAgent) ||
      /Edge/.test(window.navigator.userAgent)
    );

    /**
     * List of FlowFile objects
     * @type {Array.<FlowFile>}
     */
    this.files = [];

    /**
     * Default options for flow.js
     * @type {Object}
     */
    this.defaults = {
      chunkSize: 1024 * 1024,
      forceChunkSize: false,
      simultaneousUploads: 3,
      singleFile: false,
      fileParameterName: 'file',
      progressCallbacksInterval: 500,
      speedSmoothingFactor: 0.1,
      query: {},
      headers: {},
      withCredentials: false,
      preprocess: null,
      changeRawDataBeforeSend: null,
      method: 'multipart',
      testMethod: 'GET',
      uploadMethod: 'POST',
      prioritizeFirstAndLastChunk: false,
      allowDuplicateUploads: false,
      target: '/',
      testChunks: true,
      generateUniqueIdentifier: null,
      maxChunkRetries: 0,
      chunkRetryInterval: null,
      permanentErrors: [404, 413, 415, 500, 501],
      successStatuses: [200, 201, 202],
      onDropStopPropagation: false,
      initFileFn: null,
      readFileFn: webAPIFileRead,
      asyncReadFileFn: null
    };

    /**
     * Current options
     * @type {Object}
     */
    this.opts = {};

    /**
     * Current options
     * @type {Object}
     */
    this.opts = Object.assign({}, this.defaults, opts || {});

    // A workaround for using this.method.bind(this) as a (removable) event handler.
    // https://stackoverflow.com/questions/11565471
    this._onDropBound = null;
  }

  /**
   * On drop event
   * @function
   * @param {MouseEvent} event
   */
  onDrop(event) {
    if (this.opts.onDropStopPropagation) {
      event.stopPropagation();
    }
    event.preventDefault();
    var dataTransfer = event.dataTransfer;
    if (dataTransfer.items && dataTransfer.items[0] &&
        dataTransfer.items[0].webkitGetAsEntry) {
      this.webkitReadDataTransfer(event);
    } else {
      this.addFiles(dataTransfer.files, event);
    }
  }

  /**
   * On drop event when file/stream initialization is asynchronous
   * @function
   * @param {MouseEvent} event
   */
  async asyncOnDrop(event) {
    if (this.opts.onDropStopPropagation) {
      event.stopPropagation();
    }
    event.preventDefault();
    var dataTransfer = event.dataTransfer;
    if (dataTransfer.items && dataTransfer.items[0] &&
        dataTransfer.items[0].webkitGetAsEntry) {
      this.webkitReadDataTransfer(event);
    } else {
      await this.asyncAddFiles(dataTransfer.files, event);
    }
  }

  /**
   * Prevent default
   * @function
   * @param {MouseEvent} event
   */
  preventEvent(event) {
    event.preventDefault();
  }

  /**
   * Read webkit dataTransfer object
   * @param event
   */
  webkitReadDataTransfer(event) {
    var queue = event.dataTransfer.items.length;
    var decrement = () => {
      if (--queue == 0) {
        this.addFiles(files, event);
      }
    };

    var files = [];
    for (let item of event.dataTransfer.items) {
      var entry = item.webkitGetAsEntry();
      if (!entry) {
        decrement();
        return ;
      }
      if (entry.isFile) {
        // due to a bug in Chrome's File System API impl - #149735
        fileReadSuccess(item.getAsFile(), entry.fullPath);
      } else {
        readDirectory(entry.createReader());
      }
    }

    function readDirectory(reader) {
      reader.readEntries((entries) => {
        if (entries.length) {
          queue += entries.length;
          var fullPaths = {};
          for (let entry of entries) {
            if (entry.isFile) {
              fullPaths[entry.name] = entry.fullPath;
              entry.file((file) => fileReadSuccess(file, fullPaths[file.name]), readError);
            } else if (entry.isDirectory) {
              readDirectory(entry.createReader());
            }
          }
          readDirectory(reader);
        } else {
          decrement();
        }
      }, readError);
    }
    function fileReadSuccess(file, fullPath) {
      // relative path should not start with "/"
      file.relativePath = fullPath.substring(1);
      files.push(file);
      decrement();
    }
    function readError(fileError) {
      decrement();
      throw fileError;
    }
  }

  /**
   * Generate unique identifier for a file
   * @function
   * @param {FlowFile} file
   * @returns {string}
   */
  generateUniqueIdentifier(file) {
    var custom = this.opts.generateUniqueIdentifier;
    if (typeof custom === 'function') {
      return custom(file);
    }
    // Some confusion in different versions of Firefox
    var relativePath = file.relativePath || file.webkitRelativePath || file.fileName || file.name;
    return file.size + '-' + relativePath.replace(/[^0-9a-zA-Z_-]/img, '');
  }

  /**
   * Upload next chunk from the queue
   * @function
   * @returns {boolean}
   * @private
   */
  uploadNextChunk(preventEvents) {
    // In some cases (such as videos) it's really handy to upload the first
    // and last chunk of a file quickly; this let's the server check the file's
    // metadata and determine if there's even a point in continuing.
    var found = false;
    if (this.opts.prioritizeFirstAndLastChunk) {
      for (let file of this.files) {
        if (!file.paused && file.chunks.length &&
            file.chunks[0].status() === 'pending') {
          file.chunks[0].send();
          found = true;
          break;
        }
        if (!file.paused && file.chunks.length > 1 &&
            file.chunks[file.chunks.length - 1].status() === 'pending') {
          file.chunks[file.chunks.length - 1].send();
          found = true;
          break;
        }
      }
      if (found) {
        return found;
      }
    }

    // Now, simply look for the next, best thing to upload
    outer_loop:
    for (let file of this.files) {
      if (file.paused) {
        continue;
      }
      for (let chunk of file.chunks) {
        if (chunk.status() === 'pending') {
          chunk.send();
          found = true;
          break outer_loop;
        }
      }
    }
    if (found) {
      return true;
    }

    // The are no more outstanding chunks to upload, check if everything is done
    var outstanding = false;
    for (let file of this.files) {
      if (!file.isComplete()) {
        outstanding = true;
        break;
      }
    }

    if (!outstanding && !preventEvents) {
      // All chunks have been uploaded, complete
      this.emit('complete');
    }
    return false;
  }


  /**
   * Assign a browse action to one or more DOM nodes.
   * @function
   * @param {Element|Array.<Element>} domNodes
   * @param {boolean} isDirectory Pass in true to allow directories to
   * @param {boolean} singleFile prevent multi file upload
   * @param {Object} attributes set custom attributes:
   *  http://www.w3.org/TR/html-markup/input.file.html#input.file-attributes
   *  eg: accept: 'image/*'
   * be selected (Chrome only).
   */
  assignBrowse(domNodes, isDirectory, singleFile, attributes) {
    if (domNodes instanceof Element) {
      domNodes = [domNodes];
    }

    each(domNodes, function (domNode) {
      var input;
      if (domNode.tagName === 'INPUT' && domNode.type === 'file') {
        input = domNode;
      } else {
        input = document.createElement('input');
        input.setAttribute('type', 'file');
        // display:none - not working in opera 12
        Object.assign(input.style, {
          visibility: 'hidden',
          position: 'absolute',
          width: '1px',
          height: '1px'
        });
        // for opera 12 browser, input must be assigned to a document
        domNode.appendChild(input);
        // https://developer.mozilla.org/en/using_files_from_web_applications)
        // event listener is executed two times
        // first one - original mouse click event
        // second - input.click(), input is inside domNode
        domNode.addEventListener('click', function() {
          input.click();
        }, false);
      }
      if (!this.opts.singleFile && !singleFile) {
        input.setAttribute('multiple', 'multiple');
      }
      if (isDirectory) {
        input.setAttribute('webkitdirectory', 'webkitdirectory');
      }
      each(attributes, function (value, key) {
        input.setAttribute(key, value);
      });
      // When new files are added, simply append them to the overall list
      // but adapt to the case where initFileFn is async.
      var callback = this.opts.initFileFn
          && typeof this.opts.initFileFn === 'function'
          && this.opts.initFileFn.constructor.name  === 'AsyncFunction'
          ? async e => {
            if (e.target.value) {
              input.setAttribute('readonly', 'readonly');
              await this.asyncAddFiles(e.target.files, e);
              e.target.value = '';
              input.removeAttribute('readonly');
            }
          }
          : e => {
            if (e.target.value) {
              this.addFiles(e.target.files, e);
              e.target.value = '';
            }
          };
      input.addEventListener('change', callback, false);
    }, this);
  }

  /**
   * Assign one or more DOM nodes as a drop target.
   * @function
   * @param {Element|Array.<Element>} domNodes
   */
  assignDrop(domNodes) {
    if (typeof domNodes.length === 'undefined') {
      domNodes = [domNodes];
    }

    this._onDropBound = this.opts.initFileFn
          && typeof this.opts.initFileFn === 'function'
          && this.opts.initFileFn.constructor.name  === 'AsyncFunction'
          ? this.asyncOnDrop.bind(this)
          : this.onDrop.bind(this);
    for (let domNode of domNodes) {
      domNode.addEventListener('dragover', this.preventEvent, false);
      domNode.addEventListener('dragenter', this.preventEvent, false);
      domNode.addEventListener('drop', this._onDropBound, false);
    }
  }

  /**
   * Un-assign drop event from DOM nodes
   * @function
   * @param domNodes
   */
  unAssignDrop(domNodes) {
    if (typeof domNodes.length === 'undefined') {
      domNodes = [domNodes];
    }

    for (let domNode of domNodes) {
      domNode.removeEventListener('dragover', this.preventEvent, false);
      domNode.removeEventListener('dragenter', this.preventEvent, false);
      domNode.removeEventListener('drop', this._onDropBound, false);
    }
  }

  /**
   * Returns a boolean indicating whether or not the instance is currently
   * uploading anything.
   * @function
   * @returns {boolean}
   */
  isUploading() {
    var uploading = false;
    each(this.files, function (file) {
      if (file.isUploading()) {
        uploading = true;
        return false;
      }
    });
    return uploading;
  }

  /**
   * should upload next chunk
   * @function
   * @returns {boolean|number}
   */
  _shouldUploadNext() {
    var num = 0;
    var should = true;
    var simultaneousUploads = this.opts.simultaneousUploads;
    each(this.files, function (file) {
      each(file.chunks, function(chunk) {
        if (chunk.status() === 'uploading') {
          num++;
          if (num >= simultaneousUploads) {
            should = false;
            return false;
          }
        }
      });
    });
    // if should is true then return uploading chunks's length
    return should && num;
  }

  /**
   * Start or resume uploading.
   * @function
   */
  upload() {
    // Make sure we don't start too many uploads at once
    var ret = this._shouldUploadNext();
    if (ret === false) {
      return;
    }
    // Kick off the queue
    this.emit('upload-start');
    var started = false;
    for (var num = 1; num <= this.opts.simultaneousUploads - ret; num++) {
      started = this.uploadNextChunk(true) || started;
    }
    if (!started) {
      this.emit('complete');
    }
  }

  /**
   * Resume uploading.
   * @function
   */
  resume() {
    each(this.files, function (file) {
      if (!file.isComplete()) {
        file.resume();
      }
    });
  }

  /**
   * Pause uploading.
   * @function
   */
  pause() {
    each(this.files, function (file) {
      file.pause();
    });
  }

  /**
   * Cancel upload of all FlowFile objects and remove them from the list.
   * @function
   */
  cancel() {
    for (var i = this.files.length - 1; i >= 0; i--) {
      this.files[i].cancel();
    }
  }

  /**
   * Returns a number between 0 and 1 indicating the current upload progress
   * of all files.
   * @function
   * @returns {number}
   */
  progress() {
    var totalDone = 0;
    var totalSize = 0;
    // Resume all chunks currently being uploaded
    each(this.files, function (file) {
      totalDone += file.progress() * file.size;
      totalSize += file.size;
    });
    return totalSize > 0 ? totalDone / totalSize : 0;
  }

  /**
   * A generator to yield files and factor the sync part of the filtering logic used by both
   * addFiles & asyncAddFiles
   */
  *filterFileList(fileList, event) {
    // ie10+
    var ie10plus = window.navigator.msPointerEnabled;

    for (let file of fileList) {
      // https://github.com/flowjs/flow.js/issues/55
      if ((ie10plus && file.size === 0) || (file.size % 4096 === 0 && (file.name === '.' || file.fileName === '.'))) {
        // console.log(`file ${file.name} empty. skipping`);
        continue;
      }

      var uniqueIdentifier = this.generateUniqueIdentifier(file);
      if (!this.opts.allowDuplicateUploads && this.getFromUniqueIdentifier(uniqueIdentifier)) {
        // console.log(`file ${file.name} non-unique. skipping`);
        continue;
      }

      if (! this.hook('filter-file', file, event)) {
        // console.log(`file ${file.name} filtered-out. skipping`);
        continue;
      }

      yield [file, uniqueIdentifier];
    }
  }

  /**
   * Add a HTML5 File object to the list of files.
   * @function
   * @param {File} file
   * @param Any other parameters supported by addFiles
   */
  addFile(file, ...args) {
    this.addFiles([file], ...args);
  }

  /**
   * Add a HTML5 File object to the list of files.
   * @function
   * @param {FileList|Array} fileList
   * @param {Event} [event] event is optional
   */
  addFiles(fileList, event = null, initFileFn = this.opts.initFileFn) {
    if (this.hasHook(true)) {
      console.warn('Using addFiles() while settings asynchronous events can lead to unpredictable results. Use asyncAddFiles() instead.');
    }

    let item, file, flowfile, uniqueIdentifier, files = [];
    const iterator = this.filterFileList(fileList, event);
    while ((item = iterator.next()) && !item.done) {
      [file, uniqueIdentifier] = item.value;

      var f = new FlowFile(this, file, uniqueIdentifier);
      f.bootstrap(event, initFileFn);
      this.hook('file-added', f, event);
      this.aHook('file-added', f, event);
      if (! f.file) {
        continue;
      }
      files.push(f);
    }

    this.hook('files-added', files, event);
    this.aHook('files-added', files, event);

    files = files.filter(e => e);
    for (file of files) {
      if (this.opts.singleFile && this.files.length > 0) {
        this.removeFile(this.files[0]);
      }
      this.files.push(file);
      // console.log(`enqueue file ${file.name} of ${file.chunks.length} chunks`);
    }

    this.hook('files-submitted', this.files, event);
    this.aHook('files-submitted', this.files, event);
  }


  /**
   * Add a HTML5 File object to the list of files.
   * @function
   * @param {File} file
   * @param Any other parameters supported by asyncAddFiles.
   *
   * @return (async) An initialized <AsyncFlowFile>.
   */
  async asyncAddFile(file, ...args) {
    return (await this.asyncAddFiles([file], ...args))[0];
  }

  /**
   * Add a HTML5 File object to the list of files.
   * @function
   * @param {FileList|Array} fileList
   * @param {Event} [event] event is optional
   *
   * @return Promise{[<AsyncFlowFile>,...]} The promise of getting an array of AsyncFlowFile.
   */
  async asyncAddFiles(fileList, event = null, initFileFn = this.opts.initFileFn) {
    let item, file, flowfile, uniqueIdentifier, states = [];
    const iterator = this.filterFileList(fileList, event);

    while ((item = iterator.next()) && !item.done) {
      [file, uniqueIdentifier] = item.value;
      if (! await this.aHook('filter-file', file, event)) {
        // console.log(`file ${file.name} filtered-out. skipping`);
        continue;
      }

      // ToDo: parallelizable ?
      var flowFile = new AsyncFlowFile(this, file, uniqueIdentifier),
          state = flowFile.bootstrap(event, initFileFn);
      states.push(state);
    }

    var flowfiles = await Promise.all(states);
    for (let ff of flowfiles) {
      this.hook('file-added', ff, event);
      await this.aHook('file-added', ff, event);
    }

    this.hook('files-added', flowfiles, event);
    await this.aHook('files-added', flowfiles, event);

    flowfiles = flowfiles.filter(e => e && e.file);
    for (let file of flowfiles) {
      if (this.opts.singleFile && this.files.length > 0) {
        this.removeFile(this.files[0]);
      }
      this.files.push(file);
    }

    this.hook('files-submitted', this.files, event);
    await this.aHook('files-submitted', this.files, event);

    return flowfiles;
  }

  /**
   * Cancel upload of a specific FlowFile object from the list.
   * @function
   * @param {FlowFile} file
   */
  removeFile(file) {
    for (var i = this.files.length - 1; i >= 0; i--) {
      if (this.files[i] === file) {
        this.files.splice(i, 1);
        file.abort();
        this.emit('file-removed', file);

        if (!this.opts.allowDuplicateUploads) {
          break;
        }
      }
    }
  }

  /**
   * Look up a FlowFile object by its unique identifier.
   * @function
   * @param {string} uniqueIdentifier
   * @returns {boolean|FlowFile} false if file was not found
   */
  getFromUniqueIdentifier(uniqueIdentifier) {
    var ret = false;
    each(this.files, function (file) {
      if (file.uniqueIdentifier === uniqueIdentifier) {
        ret = file;
      }
    });
    return ret;
  }

  /**
   * Returns the total size of all files in bytes.
   * @function
   * @returns {number}
   */
  getSize() {
    var totalSize = 0;
    each(this.files, function (file) {
      totalSize += file.size;
    });
    return totalSize;
  }

  /**
   * Returns the total size uploaded of all files in bytes.
   * @function
   * @returns {number}
   */
  sizeUploaded() {
    var size = 0;
    each(this.files, function (file) {
      size += file.sizeUploaded();
    });
    return size;
  }

  /**
   * Returns remaining time to upload all files in seconds. Accuracy is based on average speed.
   * If speed is zero, time remaining will be equal to positive infinity `Number.POSITIVE_INFINITY`
   * @function
   * @returns {number}
   */
  timeRemaining() {
    var sizeDelta = 0;
    var averageSpeed = 0;
    each(this.files, function (file) {
      if (!file.paused && !file.error) {
        sizeDelta += file.size - file.sizeUploaded();
        averageSpeed += file.averageSpeed;
      }
    });
    if (sizeDelta && !averageSpeed) {
      return Number.POSITIVE_INFINITY;
    }
    if (!sizeDelta && !averageSpeed) {
      return 0;
    }
    return Math.floor(sizeDelta / averageSpeed);
  }
};
